File: backend/internal/model/submission.go
package model

import (
	"time"

	"go.mongodb.org/mongo-driver/bson/primitive"
)

type SubmissionType string

const (
	SubmissionTypeGuide SubmissionType = "guide"
	SubmissionTypeFood  SubmissionType = "food"
)

type SubmissionStatus string

const (
	SubmissionDraft             SubmissionStatus = "draft"
	SubmissionPendingReview     SubmissionStatus = "pending_review"
	SubmissionAutoRejected      SubmissionStatus = "auto_rejected"
	SubmissionNeedsManualReview SubmissionStatus = "needs_manual_review"
	SubmissionApproved          SubmissionStatus = "approved"
	SubmissionPublished         SubmissionStatus = "published"
	SubmissionRejected          SubmissionStatus = "rejected"
	SubmissionRemoved           SubmissionStatus = "removed"
)

// GuideSubmissionPayload mirrors TS GuideSubmissionPayload.
type GuideSubmissionPayload struct {
	Title       string   `bson:"title" json:"title"`
	Destination string   `bson:"destination" json:"destination"`
	Days        int      `bson:"days" json:"days"`
	Budget      float64  `bson:"budget" json:"budget"`
	Content     string   `bson:"content" json:"content"`
	Image       string   `bson:"image" json:"image"`
	Tags        []string `bson:"tags" json:"tags"`
}

// FoodSubmissionPayload mirrors TS FoodSubmissionPayload.
type FoodSubmissionPayload struct {
	Name        string   `bson:"name" json:"name"`
	Province    string   `bson:"province" json:"province"`
	City        string   `bson:"city" json:"city"`
	Address     string   `bson:"address" json:"address"`
	Type        string   `bson:"type" json:"type"`
	Price       string   `bson:"price" json:"price"`
	Description string   `bson:"description" json:"description"`
	Image       string   `bson:"image" json:"image"`
	Tags        []string `bson:"tags" json:"tags"`
}

// Submission is stored as raw BSON so polymorphic payload works without generics.
type Submission struct {
	ID                primitive.ObjectID  `bson:"_id,omitempty" json:"id"`
	Type              SubmissionType      `bson:"type" json:"type"`
	Author            string              `bson:"author" json:"author"`
	AuthorID          *primitive.ObjectID `bson:"authorId,omitempty" json:"-"`
	Payload           any                 `bson:"payload" json:"-"` // raw BSON — decoded per Type
	Status            SubmissionStatus    `bson:"status" json:"status"`
	RiskScore         float64             `bson:"riskScore" json:"riskScore"`
	RiskLabels        []string            `bson:"riskLabels" json:"riskLabels"`
	ModerationReason  string              `bson:"moderationReason,omitempty" json:"-"`
	PublishedItemID   *primitive.ObjectID `bson:"publishedItemId,omitempty" json:"-"`
	PublishedItemModel string             `bson:"publishedItemModel,omitempty" json:"-"`
	ReviewedBy        *primitive.ObjectID `bson:"reviewedBy,omitempty" json:"-"`
	ReviewedAt        *time.Time          `bson:"reviewedAt,omitempty" json:"-"`
	CreatedAt         time.Time           `bson:"createdAt" json:"createdAt"`
	UpdatedAt         time.Time           `bson:"updatedAt" json:"updatedAt"`
}

// DecodePayload unmarshals the polymorphic payload into the correct type.
func (s *Submission) DecodePayload() (any, error) {
	// The BSON driver stores Payload as bson.M / bson.D when inserted;
	// in practice we return it directly and let service code cast it.
	return s.Payload, nil
}

File: src/lib/database/models/Submission.ts
import mongoose, { type Document, type Model } from 'mongoose';

export type SubmissionType = 'guide' | 'food';

export type SubmissionStatus =
  | 'draft'
  | 'pending_review'
  | 'auto_rejected'
  | 'needs_manual_review'
  | 'approved'
  | 'published'
  | 'rejected'
  | 'removed';

export interface GuideSubmissionPayload {
  title: string;
  destination: string;
  days: number;
  budget: number;
  content: string;
  image: string;
  tags: string[];
}

export interface FoodSubmissionPayload {
  name: string;
  province: string;
  city: string;
  address: string;
  type: string;
  price: string;
  description: string;
  image: string;
  tags: string[];
}

export type SubmissionPayload = GuideSubmissionPayload | FoodSubmissionPayload;

export interface ISubmission extends Document {
  type: SubmissionType;
  author: string;
  authorId?: mongoose.Types.ObjectId;
  payload: SubmissionPayload;
  status: SubmissionStatus;
  riskScore: number;
  riskLabels: string[];
  moderationReason?: string;
  publishedItemId?: mongoose.Types.ObjectId;
  publishedItemModel?: 'Guide' | 'Food';
  reviewedBy?: mongoose.Types.ObjectId;
  reviewedAt?: Date;
  createdAt: Date;
  updatedAt: Date;
}

function isString(value: unknown): value is string {
  return typeof value === 'string' && value.trim().length > 0;
}

function isStringArray(value: unknown): value is string[] {
  return Array.isArray(value) && value.every((item) => typeof item === 'string');
}

function isValidPayload(type: SubmissionType, value: SubmissionPayload): boolean {
  if (!value || typeof value !== 'object') return false;

  if (type === 'guide') {
    const payload = value as GuideSubmissionPayload;
    return (
      isString(payload.title) &&
      isString(payload.destination) &&
      Number.isFinite(payload.days) &&
      payload.days >= 1 &&
      Number.isFinite(payload.budget) &&
      payload.budget >= 0 &&
      isString(payload.content) &&
      isString(payload.image) &&
      isStringArray(payload.tags)
    );
  }

  const payload = value as FoodSubmissionPayload;
  return (
    isString(payload.name) &&
    isString(payload.province) &&
    isString(payload.city) &&
    isString(payload.address) &&
    isString(payload.type) &&
    isString(payload.price) &&
    isString(payload.description) &&
    isString(payload.image) &&
    isStringArray(payload.tags)
  );
}

const submissionSchema = new mongoose.Schema<ISubmission>(
  {
    type: {
      type: String,
      enum: ['guide', 'food'],
      required: true,
    },
    author: { type: String, required: true, trim: true },
    authorId: { type: mongoose.Schema.Types.ObjectId, ref: 'User' },
    payload: {
      type: mongoose.Schema.Types.Mixed,
      required: true,
    },
    status: {
      type: String,
      enum: [
        'draft',
        'pending_review',
        'auto_rejected',
        'needs_manual_review',
        'approved',
        'published',
        'rejected',
        'removed',
      ],
      default: 'pending_review',
    },
    riskScore: { type: Number, default: 0 },
    riskLabels: { type: [String], default: [] },
    moderationReason: { type: String },
    publishedItemId: { type: mongoose.Schema.Types.ObjectId },
    publishedItemModel: { type: String, enum: ['Guide', 'Food'] },
    reviewedBy: { type: mongoose.Schema.Types.ObjectId, ref: 'User' },
    reviewedAt: { type: Date },
  },
  { timestamps: true }
);

submissionSchema.pre('validate', function validatePayload() {
  if (!isValidPayload(this.type, this.payload)) {
    throw new Error('Invalid submission payload.');
  }
});

submissionSchema.index({ authorId: 1, createdAt: -1 });
submissionSchema.index({ type: 1, status: 1, createdAt: -1 });

const Submission: Model<ISubmission> =
  (mongoose.models.Submission as Model<ISubmission>) ||
  mongoose.model<ISubmission>('Submission', submissionSchema);

export default Submission;
